### PR TITLE
secboot/keymgr, cmd/snap-fde-keymgr: two step encryption key change

### PR DESCRIPTION
**Pull request by bboozzoo**
Make encryption key change a 2 step process. First the new encryption key is staged (i.e. added in a temporary slot), after that, the encryption key is transitioned to become the main one. The first part of the process normally happens during factory reset, while the second in a post-reset run system.

### CODE REVIEW

File: secboot/keymgr/keymgr_luks2.go [hunk 1]
```diff
@@ -160,26 +160,24 @@ func RemoveRecoveryKeyFromLUKSDeviceUsingKey(currKey keys.EncryptionKey, dev str
...
-	// keyslot, and cannot start with killing its keyslot as that would make
-	// the device unusable, so instead add the new key to an auxiliary
-	// keyslot, then use the new key to authorize removal of keyslot 0
-	// (which refers to the old key), add the new key again, but this time
-	// to keyslot 0, lastly kill the aux keyslot
+	// TODO rather than inspecting the errors, parse the LUKS2 headers
 
+	// free up the temp slot
 	if err := luks2.KillSlot(dev, tempKeySlot, currKey); err != nil {
 		if !isKeyslotNotActive(err) {
```
Comment (pedronis) [inline secboot/keymgr/keymgr_luks2.go:182]: we don't reach here in the uni test to check this if not

File: secboot/keymgr/keymgr_luks2.go [hunk 2]
```diff
@@ -193,35 +191,86 @@ func ChangeLUKSDeviceEncryptionKey(newKey keys.EncryptionKey, dev string) error
...
+		}
+
+		return nil
+	}
+
+	// first kill the main encryption key slot, authorize the operation
+	// using the new key which must have been added to the temp keyslot in
+	// the stage operation
 	if err := luks2.KillSlot(dev, encryptionKeySlot, newKey); err != nil {
 		if !isKeyslotNotActive(err) {
```
Comment (pedronis) [inline secboot/keymgr/keymgr_luks2.go:251]: same here

File: secboot/keymgr/keymgr_luks2.go [hunk 3]
```diff
@@ -193,35 +191,86 @@ func ChangeLUKSDeviceEncryptionKey(newKey keys.EncryptionKey, dev string) error
...
-		return fmt.Errorf("cannot add key: %v", err)
+		return fmt.Errorf("cannot add new encryption key: %v", err)
 	}
-	// and kill the aux slot
+
+	// now it should be possible to kill the temporary keyslot by using the
+	// new key for authorization
 	if err := luks2.KillSlot(dev, tempKeySlot, newKey); err != nil {
-		return fmt.Errorf("cannot kill temporary key slot: %v", err)
+		if !isKeyslotNotActive(err) {
```
Comment (pedronis) [inline secboot/keymgr/keymgr_luks2.go:267]: and here

**Review by pedronis — APPROVED** [review]
lgtm, thanks for the spread test, some comment about the unit testing of the fallback case checks

File: secboot/keymgr/keymgr_luks2.go [hunk 4]
```diff
@@ -193,35 +191,86 @@ func ChangeLUKSDeviceEncryptionKey(newKey keys.EncryptionKey, dev string) error
...
+// change to the new key provided in the parameters. The new key must have been
+// staged before, thus it can authorize LUKS operations. Lastly, the unlock key
+// in the keyring is updated to the new key.
+func TransitionLUKSDeviceEncryptionKeyChange(newKey keys.EncryptionKey, dev string) error {
+	if len(newKey) != keys.EncryptionKeySize {
+		return fmt.Errorf("cannot use a key of size different than %v", keys.EncryptionKeySize)
+	}
+
+	// the expected state is as follows:
+	// keys lot 0 - the old encryption key
```
Comment (mardy) [inline secboot/keymgr/keymgr_luks2.go:207]: typo

**Pull request merged**